operators/pkg/controller/elasticsearch/driver/upgrade_test.go

```diff
@@ -0,0 +1,321 @@
+						Partition: 1,
+						Status: appsv1.StatefulSetStatus{
+							CurrentRevision: "a",
+							UpdateRevision:  "b",
+						},
+					}.Build(),
+				},
+				esState: defaultESState,
+			},
+			want: newResults(),
+			wantNewPartition: map[string]int32{
+				"default": 0,
+			},
+			wantSyncedFlush: true,
+		},
+		{
+			name: "just one (master) at a time",
+			args: args{
+				statefulSets: sset.StatefulSetList{
+					nodespec.TestSset{
+						Name:      "default",
+						Replicas:  3,
+						Master:    true,
+						Data:      true,
+						Partition: 3,
+						Status: appsv1.StatefulSetStatus{
+							CurrentRevision: "a",
+							UpdateRevision:  "b",
+						},
+					}.Build(),
+				},
+				esState: defaultESState,
+			},
+			want: newResults().WithResult(defaultRequeue),
+			wantNewPartition: map[string]int32{
+				"default": 2,
+			},
+			wantSyncedFlush: true,
+		},
+		{
+			name: "multiple ssets, update correct sset",
+			args: args{
+				statefulSets: sset.StatefulSetList{
+					nodespec.TestSset{
+						Name:     "master",
+						Replicas: 3,
+						Master:   true,
+					}.Build(),
+					nodespec.TestSset{
+						Name:      "data",
+						Replicas:  1,
+						Data:      true,
+						Partition: 1,
+						Status: appsv1.StatefulSetStatus{
+							CurrentRevision: "a",
+							UpdateRevision:  "b",
+						},
+					}.Build(),
+				},
+				esState: defaultESState,
+			},
+			want: newResults(),
+			wantNewPartition: map[string]int32{
+				"data": 0,
+			},
+			wantSyncedFlush: true,
+		},
+		{
+			name: "wait for healthy cluster",
+			args: args{
+				statefulSets: sset.StatefulSetList{
+					nodespec.TestSset{
+						Name:      "default",
+						Replicas:  1,
+						Master:    true,
+						Data:      true,
+						Partition: 2,
+						Status: appsv1.StatefulSetStatus{
+							CurrentRevision: "a",
+							UpdateRevision:  "b",
+						},
+					}.Build(),
+				},
+				esState: mockESState{
+					green: false,
+				},
+			},
+			want:             newResults().WithResult(defaultRequeue),
+			wantNewPartition: map[string]int32{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var runtimeObjects []runtime.Object
+			for i := range tt.args.statefulSets {
+				runtimeObjects = append(runtimeObjects, &tt.args.statefulSets[i])
+			}
+			k8sClient := k8s.WrapClient(fake.NewFakeClient(runtimeObjects...))
+			d := &defaultDriver{
+				DefaultDriverParameters: DefaultDriverParameters{
+					Client: k8sClient,
+					Scheme: scheme.Scheme,
+				},
+			}
+			mu := mockUpdater{}
+			fc := fakeESClient{}
+			if got := d.doRollingUpgrade(tt.args.statefulSets, &fc, tt.args.esState, mu.updatePartition); !reflect.DeepEqual(got, tt.want) {
+				result, err := got.Aggregate()
+				wantRes, wantErr := tt.want.Aggregate()
+				t.Errorf("doRollingUpgrade() = %v, want %v,err = %v, want %v", result, wantRes, err, wantErr)
+			}
+			if diff := deep.Equal(map[string]int32(mu), tt.wantNewPartition); diff != nil {
+				t.Error(diff)
```

Comment:
nit: this could be `assert.Nil(t, diff)` which might make it clearer what it's checking if it fails

Reply:
```
FAIL: Test_defaultDriver_doRollingUpgrade/multiple_ssets,_update_correct_sset (0.00s)
        upgrade_test.go:196: 
            	Error Trace:	upgrade_test.go:196
            	Error:      	Expected nil, but got: []string{"map[data]: 0 != 3"}
            	Test:       	Test_defaultDriver_doRollingUpgrade/multiple_ssets,_update_correct_sset
```
vs
```
  --- FAIL: Test_defaultDriver_doRollingUpgrade/multiple_ssets,_update_correct_sset (0.00s)
        upgrade_test.go:196: [map[data]: 0 != 3]
```

Reply:
Not sure which is better, second one without `assert.Nil` is more to the point but maybe lacks some context?

Reply:
Nah that makes sense to me, I retract my nit